<a href="https://colab.research.google.com/github/tlvhoang06/midterm-lab-combinatorics/blob/main/MidTerm_Lab_ipynb_e593e0d0e25e47567b593818d4936445_txt.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# BỘ ĐỀ ÔN LUYỆN THỰC HÀNH MÔN TOÁN HỌC TỔ HỢP

## Cấu trúc chung

Đây là bộ đề ôn luyện thi thực hành hoàn chỉnh gồm 20 câu. Sử dụng các kỹ thuật lập trình Python đã học (sử dụng vòng lặp có điều kiện, các hàm $math$, $itertools$, và $sympy.FiniteSet$).

Mỗi câu trả lời được trình bày theo cấu trúc 3 phần yêu cầu:
- Code đếm bằng vòng lặp/hàm Python cơ bản để liệt kê và tính toán ra lời giải,
- Lời giải bằng Toán có giải thích,
- Code so sánh kết quả.


Gợi ý: Phần code sử dụng itertools.permutations, itertools.combinations, ... hoặc for loop để tạo liệt kê. Sử dụng math.factorial, math.comb ... để tính giá trị công thức như đã học trên lớp thực hành

## Chủ đề 1: Các quy tắc đếm cơ bản

### Câu ví dụ: Đếm bội số

### Có bao nhiêu bội số của 7 trong các số nguyên từ 1 đến 1001?

#### Lời giải ví dụ:

##### 1. Code đếm lời giải (Vòng lặp có điều kiện)

In [658]:
N = 1001
divisor = 7
count_loop = 0
for i in range(1, N + 1):
    if i % divisor == 0:
        count_loop += 1
print(f"Code đếm bằng vòng lặp: {count_loop}")


Code đếm bằng vòng lặp: 143


##### 2. Lời giải bằng Toán, có giải thích

Đây là bài toán đếm số phần tử trong một dãy số chia hết cho k

- Công thức: Số bội số của k trong khoảng $[1,N]$ là $\lfloor \frac{n}{k} \rfloor$.
- Giải thích: Ta cần tìm số nguyên sao $m$ cho $7m <= 1001$

$$ m = \lfloor \frac{1001}{7} \rfloor = 143$$

##### 3. Code so sánh lời giải

In [659]:
import math
N = 1001
divisor = 7
count_math = math.floor(N / divisor)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")


Code đếm bằng vòng lặp: 143
Code sử dụng công thức: 143
Kết quả so sánh: True


### Câu 1:

### Có bao nhiêu số chẵn từ 100 đến 999 có các chữ số đôi một khác nhau?

## Giải bằng toán học
####**+ Gọi "ABC" là số từ 100 đến 999**
 - Cách chọn A : 1 -> 9
 - Cách chọn B : 1 -> 9
 - Cách chọn C : 0, 2, 4, 6, 8

####**+ TH1: C = 0**
 - C : 1 cách chọn
 - B : 9 cách chọn
- A : 8 cách chọn
 #### --> Có 1 x 9 x 8 = 72 cách
#### **+ TH2: C != 0**
- C : 4 cách chọn
 - A : 8 cách chọn
 - B : 8 cách chọn
 --> có 4 x 8 x 8 = 256 cách
#### **==> Vậy có tổng cộng 256 + 72 = 328 cách chọn**

In [660]:
count_loop = 0
for i in range(100, 999):
  if i % 2 == 0:
    s = str(i)
    if(s[0] != s[1] and s[1] != s[2] and s[0] != s[2]):
      count_loop = count_loop + 1

In [661]:
count_math = 1*9*8 + 4*8*8
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 328
Code sử dụng công thức: 328
Kết quả so sánh: True


### Câu 2:

### Bao nhiêu số lẻ 5 chữ số có chính xác một cặp chữ số trùng (tất cả chữ số khác đều khác nhau), và chữ số cuối là lẻ?

## **Giải bằng toán học**
#### Gọi số cần tìm là abcde
#### Ràng buộc:
- e có 5 cách chọn: 1, 3, 5, 7, 9
- a có 9 cách chọn: 1 -> 9
####Xét: Vị trí trùng nằm ở e
TH1: a = e  
 - a và e có 5 cách chọn (thỏa a != 0)  
 - 3 vị trí còn lại có 9P3 cách chọn  
####-> Có 5 * 9P3 = 2520 (Cách)  
TH2: a != e  
 - e và 1 vị trí i nào đó có 5 cách chọn   
 - a có 8 cách chọn
 - 2 vị trí còn lại có 8P2 cách chọn
####-> Có 3 * 5 * 8 * 8P2 = 6720 (Cách)  
### Vậy vị trí trùng ở e có tổng cộng 2520 + 6720 = 9240 (Cách)
####Xét: Vị trí trùng không nằm ở e
TH1: nằm ở a
- e có 5 cách chọn
- a và vị trí trùng còn lại có 8 cách chọn
- 2 vị trí còn lại có 8P2 cách chọn
####-> Có 3 * 5 * 8 * 8P2 = 6720 (Cách)  
TH2: không nằm ở a  
TH2.1: giá trị trùng != 0  
  - e có 5 cách chọn
  - cặp trùng có 8 cách chọn
  - a có 7 cách chọn
  - vị trí còn lại có 7 cách chọn  
TH2.2: giá trị trùng = 0  
  - e có 5 cách chọn
  - cặp trùng 1 cách chọn
  - a có 8 cách chọn
  - vị trí còn lại có 7 cách chọn
####-> 3 * 5 * 8 * 7 * 7 + 3 * 5 * 1 * 8 * 7 = 6720 (Cách)  
### Vậy vị trí trùng không nằm ở e có tổng cộng 6720 + 6720 = 13440 (Cách)

## Tổng: 9240 + 13440 = 22680 (Cách)



In [662]:
count_loop = 0
for i in range(10000, 100000):
    if i % 2 == 0:
        continue
    s = str(i)
    valid = True
    pair = 0
    for j in range(len(s)):
        freq = s.count(s[j])
        if freq > 2:
            valid = False
            break
        if freq == 2:
             pair += 1
    if valid and pair == 2:
        count_loop += 1

In [663]:
count_math = 9240 + 13440
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 22680
Code sử dụng công thức: 22680
Kết quả so sánh: True


## Chủ đề 2: Hoán vị, Chỉnh hợp, Tổ hợp


### Câu 3:
### Từ các ký tự $ {a,b,c,d,e,f} $, có bao nhiêu cách tạo ra một chuỗi gồm 3 ký tự tùy ý?

## Giải bằng toán học
### Có 3 vị trí, mỗi vị trí có 6 cách chọn ký tự
###-> 6^3 cách

In [664]:
from itertools import permutations
chars = ['a', 'b', 'c', 'd', 'e', 'f']
count_loop = 0
for a in chars:
    for b in chars:
        for c in chars:
            count_loop += 1

In [665]:
count_math = pow(6, 3)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 216
Code sử dụng công thức: 216
Kết quả so sánh: True


### Câu 4:

### Từ các ký tự $ {a,b,c,d,e,f} $, có bao nhiêu cách tạo ra một chuỗi gồm 3 ký tự đôi một khác nhau??

## Giải bằng toán học
### Chọn 3 ký tự từ 6 ký tự, có thể hoán đổi -> dùng chỉnh hợp
###-> 6P3 cách

In [666]:
from itertools import permutations
chars = ['a', 'b', 'c', 'd', 'e', 'f']
count_loop = len(list(permutations(chars, 3)))

In [667]:
from math import perm
count_math = perm(len(chars), 3)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 120
Code sử dụng công thức: 120
Kết quả so sánh: True


### Câu 5:

### Một câu lạc bộ cờ vua có 12 thành viên. Có bao nhiêu cách chọn ra một đội gồm 5 người để tham gia giải đấu (không quan tâm thứ tự)?

## Giải bằng toán học
### Chọn 5 người từ 12 người không quan tâm thứ tự -> dùng tổ hợp
###-> 12C5 cách

In [668]:
from itertools import combinations
members = range(12)
count_loop = len(list(combinations(members, 5)))

In [669]:
from math import comb
count_math = comb(len(members), 5)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 792
Code sử dụng công thức: 792
Kết quả so sánh: True


### Câu 6:

### Có bao nhiêu chuỗi nhị phân chiều dài 8 chứa đúng 3 bit 0?

## Giải bằng toán học
### Bài toán được hiểu đơn giản là chọn 3 ví trị gắn bit 0 trong 8 vị trí
###-> 8C3 chuỗi

In [670]:
from itertools import combinations
bits = range(8)
count_loop = len(list(combinations(bits, 3)))

In [671]:
from math import comb
count_math = comb(len(bits), 3)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 56
Code sử dụng công thức: 56
Kết quả so sánh: True


### Câu 7:

### Trong một giải chạy marathon có 20 vận động viên tham gia. Có bao nhiêu khả năng khác nhau cho 3 vị trí Nhất, Nhì, và Ba?

## Giải bằng toán học
### Bài toán được hiểu đơn giản là chọn 3 ví trị nhất nhì ba (có thứ tự) từ 20 vận động viên
###-> 20P3 khả năng

In [672]:
from itertools import permutations
athletes = range(20)
count_loop = len(list(permutations(athletes, 3)))

In [673]:
from math import perm
count_math = perm(len(athletes), 3)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 6840
Code sử dụng công thức: 6840
Kết quả so sánh: True


### Câu 8:

### Từ một nhóm gồm 8 nhà khoa học nam và 6 nhà khoa học nữ, có bao nhiêu cách thành lập một ủy ban gồm 6 người, sao cho ủy ban có ít nhất 3 nhà khoa học nữ?

## Giải bằng toán học
### Ta chia thành các trường hợp theo số lượng nữ trong ủy ban:
### TH1: Có 3 nữ -> 6C3 * 8C3
### TH2: Có 4 nữ -> 6C4 * 8C2
### TH3: Có 5 nữ -> 6C5 * 8C1
### TH4: Có 6 nữ -> 6C6 * 8C0
### -> Cộng tất cả các trường hợp lại được 1589 cách



In [674]:
from itertools import combinations
count_loop = 0
scientists = ['M'] * 8 + ['F'] * 6
for combo in combinations(scientists, 6):
  if combo.count('F') >= 3:
    count_loop += 1

In [675]:
from math import comb
count_math = comb(6, 3) * comb(8, 3) + comb(6, 4) * comb(8, 2) + comb(6, 5) * comb(8, 1) + comb(6, 6) * comb(8, 0)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 1589
Code sử dụng công thức: 1589
Kết quả so sánh: True


### Câu 9:

### Có bao nhiêu cách sắp xếp 8 ký tự của từ VISITING? (I lặp 3 lần, V, S, T, N, G lặp 1 lần)


## Giải bằng toán học
### Chọn 3 vị trí từ 8 vị trí cho 3 chữ I (không có thứ tự): 8C3 (cách)
### Sắp xếp 5 chữ còn lại vào 5 vị trí: 5! cách
## Vậy tổng cộng có 8C3 * 5! = 6720 (cách sắp xếp)



In [676]:
from itertools import permutations
String = "VISITING"
unique = set(permutations(String))
count_loop = len(unique)

In [677]:
from math import comb
from math import factorial
count_math = comb(8, 3) * factorial(5)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 6720
Code sử dụng công thức: 6720
Kết quả so sánh: True


### Câu 10:

### Có bao nhiêu cách sắp xếp 6 quyển sách Toán khác nhau và 4 quyển sách Hóa học khác nhau lên một kệ sách dài nếu các sách cùng một môn phải được xếp cạnh nhau?

## Giải bằng toán học
### Vì cùng môn phải xếp cạnh nhau nên ta tách ra thành 2 cụm sách Toán và Hóa: 2!
### Sắp xếp các quyển sách Toán: 6!
### Sắp xếp các quyển sách Hóa: 4!
## Vậy tổng cộng có 2! * 4! * 6! = 34560 (cách sắp xếp)



In [678]:
import itertools

toan = [f"T{i}" for i in range(1, 7)]
hoa = [f"H{i}" for i in range(1, 5)]

list_toan = list(itertools.permutations(toan))
list_hoa = list(itertools.permutations(hoa))

res = []

for i in ("TH", "HT"):
    for t in list_toan:
        for h in list_hoa:
            if i == "TH":
              res.append(t + h)
            else:
              res.append(h + t)

count_loop = len(res)

In [679]:
from math import factorial
count_math = factorial(2) * factorial(4) * factorial(6)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 34560
Code sử dụng công thức: 34560
Kết quả so sánh: True


### Câu 11:

### Một người đi mua 12 quả trái cây từ một cửa hàng có bán 5 loại: táo, cam, chuối, lê, và xoài. Hỏi có bao nhiêu cách chọn khác nhau? (Giả sử cửa hàng có đủ số lượng mỗi loại).

## Giải bằng toán học
### Dùng tổ hợp có lặp với n = 12, k = 5: (n + k - 1)C(k-1) = 16C4 = 1820

In [680]:
from itertools import combinations_with_replacement
n = 12
k = 5
fruits = ['T', 'C', 'Ch', 'L', 'X']
count_loop = len(list(combinations_with_replacement(fruits, n)))

In [681]:
from math import comb
count_math = comb(n + k - 1, k - 1)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 1820
Code sử dụng công thức: 1820
Kết quả so sánh: True


### Câu 12:

### Có bao nhiêu cách chia 10 lá thư giống hệt nhau vào 4 hộp thư khác nhau?

## Giải bằng toán học
### Dùng tổ hợp lặp với n = 10, k = 4: (n + k - 1)C(k-1) = 13C3 = 286

In [682]:
from itertools import combinations_with_replacement
n = 10
k = 4
post = range(4)
count_loop = len(list(combinations_with_replacement(post, n)))

In [683]:
from math import comb
count_math = comb(n + k - 1, k - 1)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 286
Code sử dụng công thức: 286
Kết quả so sánh: True


### Câu 13:

### 5 cặp vợ chồng đi dự đám cưới được xếp ngồi vào một bàn tròn sao cho các ông chồng ngồi cạnh nhau và các bà vợ ngồi cạnh nhau.

## Giải bằng toán học
### Số cách sắp xếp 5 ông chồng: 5!
### Số cách sắp xếp 5 bà vợ: 5!
### 2 cụm người trên bàn tròn : (2-1)!
## Vậy có: 5! * 5! * (2-1)!  = 14400

In [684]:
from itertools import permutations
chong = [f"c{i}" for i in range(1, 6)]
vo = [f"v{i}" for i in range(1, 6)]
list_chong = list(permutations(chong))
list_vo = list(permutations(vo))
res = []
for t in list_chong:
    for h in list_vo:
        res.append(t + h)
count_loop = len(res)

In [685]:
from math import factorial
count_math = factorial(5) * factorial(5)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 14400
Code sử dụng công thức: 14400
Kết quả so sánh: True


### Câu 14:

### Có bao nhiêu cách chọn ít nhất một trái từ 6 trái táo (giống nhau) và 8 trái cam (giống nhau)?

## Giải bằng toán học
### Số cách chọn táo: 0 -> 6 = 7 cách
### Số cách chọn cam: 0 -> 8 = 9 cách
### Trừ trường hợp không lấy trái nào từ 2 loại trái trên, ta được kết quả:
### 7 * 9 - 1 = 62

In [686]:
count_loop = 0
tao = 6
cam = 8
for t in range(tao + 1):
  for c in range(cam + 1):
    if(t + c > 0):
      count_loop += 1

In [687]:
count_math = 7 * 9 - 1
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 62
Code sử dụng công thức: 62
Kết quả so sánh: True


### Câu 15:

### Một tiệm kem có 10 loại kem khác nhau. Một khách hàng muốn mua 3 loại kem khác nhau trong đó phải có kem dừa.

## Giải bằng toán
### Loại kem 1: Kem dừa (bắt buộc): 1 cách
### 2 loại kem còn lại: 9C2
## Vậy có 1 * 9C2 = 36 cách

In [688]:
from itertools import combinations
n = 10
k = 3
count_loop = 0
# assume coconut ice cream is 0
ice_cream = range(n)
list_comb = list(combinations(ice_cream, k))
for combo in list_comb:
  if 0 in combo:
    count_loop += 1

In [689]:
from math import comb
count_math = comb(n - 1, k - 1)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 36
Code sử dụng công thức: 36
Kết quả so sánh: True


## Chủ đề 3: Nguyên lý Bù trừ

### Câu 16:

### Có bao nhiêu số nguyên từ 1 đến 100 chia hết cho 2, 3 hoặc 5?

## Giải bằng toán học
### Số nguyên từ 1 đến 100 chia hết cho 2: floor(100 / 2) = 50
### Số nguyên từ 1 đến 100 chia hết cho 2: floor(100 / 3) = 33
### Số nguyên từ 1 đến 100 chia hết cho 2: floor(100 / 5) = 20
### Số nguyên từ 1 đến 100 chia hết cho 2 và 5: floor(100 / LCM(2, 5)) = 10
### Số nguyên từ 1 đến 100 chia hết cho 2 và 3: floor(100 / LCM(2,3)) = 16
### Số nguyên từ 1 đến 100 chia hết cho 3 và 5: floor(100 / LCM(3,5)) = 6
### Số nguyên từ 1 đến 100 chia hết cho 2, 3 và 5: floor(100 / LCM(2, 3, 5)) = 3
## Áp dụng nguyên lý bù trừ: 50 + 33 + 20 - 10 - 16 - 6 + 3 = 74



In [690]:
count_loop = 0
for i in range(1, 101):
  if i % 2 == 0 or i % 3 == 0 or i % 5 == 0:
    count_loop += 1

In [691]:
count_math = 50 + 33 + 20 - 10 - 16 - 6 + 3
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 74
Code sử dụng công thức: 74
Kết quả so sánh: True


### Câu 17:

### Một hiệu bánh có 30 loại bánh khác nhau (18 mặn, 12 ngọt). Một khách hàng muốn chọn 5 loại bánh khác nhau sao cho đội hình gồm cả 2 loại mặn và ngọt?

## Giải bằng toán học
### Chọn 5 loại bánh toàn mặn: 18C5
### Chọn 5 loại bánh toàn ngọt: 12C5
### Chọn ngẫu nhiên: 30C5
## Vậy đội hình gồm cả 2 loại mặn ngọt: 30C5 - 18C5 - 12C5

In [692]:
from itertools import combinations
cake = ["salty"] * 18 + ["sweet"] * 12
list_comb = list(combinations(cake, 5))
count_loop = 0
for combo in list_comb:
  if "salty" in combo and "sweet" in combo:
    count_loop += 1

In [693]:
from math import comb
count_math = comb(30, 5) - comb(18, 5) - comb(12, 5)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 133146
Code sử dụng công thức: 133146
Kết quả so sánh: True


### Câu 18:

### Mã PIN gồm 4 chữ số (0-9). Mã PIN hợp lệ phải có ít nhất một chữ số lớn hơn 5 (≥6) và ít nhất một chữ số nhỏ hơn hoặc bằng 5 (≤5). Hỏi có bao nhiêu mã PIN hợp lệ?

## Giải bằng toán học
### Mã pin gồm 4 chữ số: 10^4 cách
### Mã pin không có chữ số lớn hơn 5: có 6^4 cách
### Mã pin không có chữ số nào bé hơn hoặc bằng 5: 4^4 cách
## Áp dụng nguyên lý bù trừ: 10^4 - 6^4 - 4^4 = 8448


In [694]:
from itertools import product
count_loop = 0
PIN = range(0, 10)
list_comb = list(product(PIN, repeat = 4))
for comb in list_comb:
  if any(x > 5 for x in comb) and any(x <= 5 for x in comb):
    count_loop += 1
count_loop

8448

In [695]:
count_math = pow(10, 4) - pow(6, 4) - pow(4, 4)
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 8448
Code sử dụng công thức: 8448
Kết quả so sánh: True


### Câu 19:

### Một nhân viên kho có 6 nhãn vận chuyển cho 6 thùng hàng riêng biệt. Anh ta dán nhãn một cách ngẫu nhiên. Có bao nhiêu cách dán nhãn để ít nhất một thùng hàng được dán đúng nhãn của nó?

## Giải bằng toán học
### Dán nhãn ngẫu nhiên: 6P6 = 720 cách
### 1 đúng 5 sai: 5! * 6C1
### 2 đúng 4 sai: 4! * 6C2
### 3 đúng 3 sai: 3! * 6C3
### 4 đúng 2 sai: 2! * 6C4
### 5 đúng 1 sai: 1! * 6C5
### 6 đúng 0 sai: 0! * 6C6
## Áp dụng nguyên lý bù trừ: Ta được


In [696]:
from itertools import permutations
count_loop = 0
boxes = range(1, 7)
list_perm = list(permutations(boxes))
for perm in list_perm:
  if any(perm[i] == boxes[i] for i in range(6)) in perm:
    count_loop += 1
count_loop

455

In [697]:
import math

n = 6
op = 0
for k in range(0, n+1):
    op += ((-1)**k) * math.comb(n, k) * math.factorial(n-k)

count_math = math.factorial(n) - op
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 455
Code sử dụng công thức: 455
Kết quả so sánh: True


### Câu 20:

### Trong các số (1..200), có bao nhiêu số chia hết cho 3 hoặc 7 hoặc 9?

## Giải bằng toán học
### Chia hết cho 3: floor(200 / 3) = 66
### Chia hết cho 7: floor(200 / 7) = 28
### Chia hết cho 9: floor(200 / 9) = 22
### Chia hết cho 3 và 7: floor(200 / LCM(3, 7)) = 9
### Chia hết cho 3 và 9: floor(200 / LCM(3,9)) = 22
### Chia hết cho 7 và 9: floor(200 / LCM(7, 9)) = 3
### Chia hết cho 3, 7 và 9: floor(200 / LCM(3, 7, 9)) = 3
## Áp dụng nguyên lý bù trừ: 66 + 28 + 22 - 9 - 22 - 3 + 3 = 85


In [698]:
count_loop = 0
for i in range(1, 201):
  if (i % 3 == 0) or (i % 7 == 0) or (i % 9 == 0):
    count_loop += 1

In [699]:
count_math = 66 + 28 + 22 - 9 - 22 - 3 + 3
print(f"Code đếm bằng vòng lặp: {count_loop}")
print(f"Code sử dụng công thức: {count_math}")
print(f"Kết quả so sánh: {count_loop == count_math}")

Code đếm bằng vòng lặp: 85
Code sử dụng công thức: 85
Kết quả so sánh: True
